src/dmd/dinifile.d

```diff
@@ -144,20 +144,13 @@ void updateRealEnvironment(StringTable* environment)
 {
     static int envput(const(StringValue)* sv)
     {
-        const name = sv.toDchars();
-        const namelen = strlen(name);
-        const value = cast(const(char)*)sv.ptrvalue;
+        const name = sv.toString();
+        const value = (cast(const(char)*)sv.ptrvalue)[0 .. sv.length];
         if (!value) // deleted?
             return 0;
-        const valuelen = strlen(value);
-        auto s = cast(char*)malloc(namelen + 1 + valuelen + 1);
-        assert(s);
-        memcpy(s, name, namelen);
-        s[namelen] = '=';
-        memcpy(s + namelen + 1, value, valuelen);
-        s[namelen + 1 + valuelen] = 0;
+        const s = name ~ '=' ~ value ~ '\0';
```

Comment:
You are replacing a malloc'ed string with a GC-ed string. My guess is that this will get collected by the GC improperly.

Reply:
Yeah, looks like that's the case. It looks like some implementations (including glibc) do not make copies and expect the original pointer to remain valid... eugh.